packages/itwin/tree-widget/src/components/trees/models-tree/internal/QueryHandler.ts

```diff
@@ -0,0 +1,457 @@
+                      },
+                      direction: "Forward",
+                      count: "*",
+                    },
+                  ],
+                },
+              ],
+            },
+          ],
+        },
+        keys: new KeySet([node]),
+      }),
+    ).pipe(
+      mergeMap((x) => x.items()),
+      map((x) => x.id),
+      shareReplay(),
+    );
+
+    obs = elementIds.pipe(
+      first(),
+      mergeMap((id) => {
+        const bindings = new Array<string>();
+        const query = `
+          SELECT Model.Id AS modelId, Category.Id AS categoryId
+          FROM bis.GeometricElement3d
+          WHERE ${bind("ECInstanceId", id, bindings)}
+          LIMIT 1
+        `;
+        return from(this.runQuery(query, bindings, (row) => ({ modelId: row.modelId, categoryId: row.categoryId, elementIds })));
+      }),
+      shareReplay(),
+    );
+    this._groupedElementIdsCache.set(cacheKey, obs);
+    return obs;
+  }
+
+  public queryElements(props: ElementsQueryProps): Observable<string> {
+    return this.runElementsQuery({
+      ...props,
+      type: "ids",
+    });
+  }
+
+  public queryElementsCount(props: ElementsQueryProps): Observable<number> {
+    return this.runElementsQuery({
+      ...props,
+      type: "count",
+    });
+  }
+
+  private runElementsQuery(
+    props: ElementsQueryProps & {
+      type: "ids";
+    },
+  ): Observable<Id64String>;
+  private runElementsQuery(
+    props: ElementsQueryProps & {
+      type: "count";
+    },
+  ): Observable<number>;
+  private runElementsQuery(
+    props: ElementsQueryProps & {
+      type: "ids" | "count";
+    },
+  ): Observable<Id64String | number> {
+    const bindings = new Array<QueryBindable>();
+    const conditions = new Array<string>();
+    props.modelId && conditions.push(bind("Model.Id", props.modelId, bindings));
+    props.categoryId && conditions.push(bind("Category.Id", props.categoryId, bindings));
+    props.rootElementId && conditions.push(bind("Parent.Id", props.rootElementId, bindings));
+
+    const whereClause = conditions.length ? `WHERE ${conditions.join(" AND ")}` : "";
+
+    const query = /* sql */ `
+      WITH RECURSIVE
+        RootElements(id) AS (
+          SELECT e.ECInstanceId AS id
+          FROM bis.GeometricElement3d e
+          ${whereClause}
+        ),
+        ChildElements(id) AS (
+          SELECT * FROM RootElements
+
+          UNION ALL
+
+          SELECT e.ECInstanceId AS id
+          FROM bis.GeometricElement3d e
+          INNER JOIN ChildElements ce ON e.Parent.Id = ce.id
+        )
+        SELECT ${props.type === "ids" ? "*" : "COUNT(*)"} FROM ChildElements
+        ${props.elementIds ? `WHERE ${bind("id", props.elementIds, bindings)}` : ""}
+    `;
+    return this.runQuery(query, bindings, (row) => row[0]);
+  }
+
+  private runQuery<TResult>(
+    query: string,
+    bindings: Array<QueryBindable> | Map<string, Id64String> | undefined,
```

Comment:
In that case, we'd have to write `createBinder(bindings)` in all methods that use this. Why not do it here?

Reply:
Let's take the above `runElementsQuery` function as an example. Basically, you have this:

```ts
// create the bindings array
const bindings = new Array<QueryBindable>();
// add items to `bindings` through the `bind` function like this:
bind("Model.Id", props.modelId, bindings)
// call run query, which takes the array and converts it to `QueryBinder`
return this.runQuery(query, bindings, (row) => row[0]);
```

I think it could be changed to:

```ts
// create the `QueryBinder`
const bindings = new QueryBinder();
// now, the `bind` function binds using `QueryBinder` directly
bind("Model.Id", props.modelId, bindings)
// call run query with `QueryBinder`, no more need to call `createBinder` there
return this.runQuery(query, bindings, (row) => row[0]);
```

Am I missing something?

Reply:
We must specify an index or a string key to the `QueryBinder`.
Having an array helps us to keep the index unique for each `bind` call.
I reused this logic from the `unified-selection` package, where it's done in a similar way.

Reply:
Okay. I see you also removed binding by string key, which simplifies things a bit and I think we can keep it. Thanks.